components/core/src/clp/aws/AwsAuthenticationSigner.cpp

```diff
@@ -0,0 +1,337 @@
+ * @param scope
+ * @param timestamp_string
+ * @param canonical_request
+ * @param string_to_sign Returns the string to sign
+ * @return ErrorCode_Success on success
+ * @return Same as `get_sha256_hash` on failure.
+ */
+[[nodiscard]] auto get_string_to_sign(
+        string_view scope,
+        string_view timestamp_string,
+        string_view canonical_request,
+        string& string_to_sign
+) -> ErrorCode {
+    vector<unsigned char> signed_canonical_request;
+    if (auto error_code = get_sha256_hash(
+                {size_checked_pointer_cast<unsigned char const>(canonical_request.data()),
+                 canonical_request.size()},
+                signed_canonical_request
+        );
+        ErrorCode_Success != error_code)
+    {
+        return error_code;
+    }
+    auto const signed_canonical_request_str = convert_to_hex_string(
+            {signed_canonical_request.data(), signed_canonical_request.size()}
+    );
+    string_to_sign = fmt::format(
+            "{}\n{}\n{}\n{}",
+            cAws4HmacSha256,
+            timestamp_string,
+            scope,
+            signed_canonical_request_str
+    );
+    return ErrorCode_Success;
+}
+
+/**
+ * Encodes the Uri as specified by AWS Signature Version 4's UriEncode()
+ * @param Uri
+ * @param encode_slash
+ * @return The encoded Uri
+ */
+[[nodiscard]] auto encode_uri(string_view uri, bool encode_slash) -> string {
+    string encoded_uri;
+
+    for (auto const c : uri) {
+        if ((std::isalnum(c) != 0) || c == '-' || c == '_' || c == '.' || c == '~') {
+            encoded_uri += c;
+        } else if (c == '/' && false == encode_slash) {
+            encoded_uri += c;
+        } else {
+            encoded_uri += fmt::format("%{:02X}", c);
+        }
+    }
+
+    return encoded_uri;
+}
+
+/**
+ * Gets the scope specified by AWS Signature Version 4 format
+ * @param date_string
+ * @param region
+ * @return The scope as a string
+ */
+[[nodiscard]] auto get_scope(string_view date_string, string_view region) -> string {
+    return fmt::format("{}/{}/{}/{}", date_string, region, cS3Service, cAws4Request);
+}
+
+/**
+ * Gets the canonical request string
+ * @param method HTTP method
+ * @param url S3 Url
+ * @param query_string Query string
+ * @return Canonical request
+ */
+[[nodiscard]] auto get_canonical_request(S3Url const& url, string_view query_string) -> string {
+    return fmt::format(
+            "{}\n{}\n{}\n{}:{}\n\n{}\n{}",
+            get_method_string(AwsAuthenticationSigner::HttpMethod::GET),
+            encode_uri(url.get_path(), false),
+            query_string,
+            cDefaultSignedHeaders,
+            url.get_host(),
+            cDefaultSignedHeaders,
+            cUnsignedPayload
+    );
+}
+}  // namespace
+
+S3Url::S3Url(string const& url) {
+    // Virtual-hosted-style HTTP URL format
+    std::regex const host_style_url_regex(
+            R"(https://([a-z0-9.-]+)\.s3(\.([a-z0-9-]+))?\.amazonaws\.com(/[^?]+).*)"
```

Comment:
To implement a general S3 URL parser, it might also be worth to parse the endpoint: it won't always be `s3.[REGION].amazonaws.com`. For example, BOS also provides S3 APIs such as `https://s3.bj.bcebos.com`, where the endpoint `bcebos.com`. The parsing/signing logic should be the same for BOS.